### PR TITLE
Package as a Docker image

### PR DESCRIPTION
**Pull request by sobolevn**
Related to https://github.com/thechangelog/changelog.com/issues/120

Steps to run it:
1. `docker-compose run web mix do ecto.create, ecto.migrate` to setup a database (this could be moved into `Dockerfile`
2. `docker-compose up`

I would like to discuss a few things here:
- No initial data. This is an issue because it raises a lot of exceptions. Like:

> web_1  | [debug] *\* (Ecto.NoResultsError) expected at least one result but got none in query:
> web_1  | 
> web_1  | from p in Changelog.Podcast,
> web_1  |   where: p.slug == ^"js"
> web_1  | 
> web_1  |     (ecto) lib/ecto/repo/queryable.ex:74: Ecto.Repo.Queryable.one!/4
> web_1  |     (changelog) web/controllers/episode_controller.ex:9: Changelog.EpisodeController.show/2
> web_1  |     (changelog) web/controllers/episode_controller.ex:1: Changelog.EpisodeController.action/2
> web_1  |     (changelog) web/controllers/episode_controller.ex:1: Changelog.EpisodeController.phoenix_controller_pipeline/2
> web_1  |     (changelog) lib/changelog/endpoint.ex:1: Changelog.Endpoint.instrument/4
> web_1  |     (changelog) lib/phoenix/router.ex:261: Changelog.Router.dispatch/2
> web_1  |     (changelog) web/router.ex:1: Changelog.Router.do_call/2
> web_1  |     (changelog) lib/changelog/endpoint.ex:1: Changelog.Endpoint.phoenix_pipeline/1
> web_1  |     (changelog) lib/plug/debugger.ex:123: Changelog.Endpoint."call (overridable 3)"/2
> web_1  |     (changelog) lib/changelog/endpoint.ex:1: Changelog.Endpoint.call/2
> web_1  |     (plug) lib/plug/adapters/cowboy/handler.ex:15: Plug.Adapters.Cowboy.Handler.upgrade/4
> web_1  |     (cowboy) src/cowboy_protocol.erl:442: :cowboy_protocol.execute/4
> web_1  | 
- Setting database URL, there are different options. Which one do you like?
- Strict versions. Would you like your container versions to be strict (including `node` version in `web` container)?
- Documentation.

### CODE REVIEW

**Comment by jerodsanto** [comment]
@sobolevn Awesome!

1. Seeds – We definitely need some seed data. Probably The Changelog, Go Time, a published episode of each, and an admin Person. We can use `seeds.exs` for this and then add `mix run priv/repo/seeds.exs` to the setup process. Let me know if you want to tackle that or if you'd prefer I do it.
2. DB URL – My only preference is that it won't affect people who are not using Docker. 
3. Strict versions – tell me what you think is best
4. Docs – Yes for sure. I was thinking readme at least. What else do you have in mind?

**Comment by gerhard** [comment]
@sobolevn I am wondering whether we could learn from how this is currently done in production.

This is the Dockerfile that we are currently using to create the app image:

```Docker
FROM erlang:18.3.4.1

# elixir expects utf8.
ENV ELIXIR_VERSION="v1.3.2" \
  LANG=C.UTF-8

RUN set -xe \
  && ELIXIR_DOWNLOAD_URL="https://github.com/elixir-lang/elixir/releases/download/${ELIXIR_VERSION}/Precompiled.zip" \
  && ELIXIR_DOWNLOAD_SHA256="45fdb9464b0fbe44c919482f1247740cc9c5d399280ef07e386aa7402b085be7"\
  && buildDeps=' \
    unzip \
  ' \
  && apt-get update \
  && apt-get install -y --no-install-recommends $buildDeps \
  && curl -fSL -o elixir-precompiled.zip $ELIXIR_DOWNLOAD_URL \
  && echo "$ELIXIR_DOWNLOAD_SHA256 elixir-precompiled.zip" | sha256sum -c - \
  && unzip -d /usr/local elixir-precompiled.zip \
  && rm elixir-precompiled.zip \
  && apt-get purge -y --auto-remove $buildDeps \
  && rm -rf /var/lib/apt/lists/*

RUN apt-get update \
    && apt-get install -y apt-utils \
    && rm -rf /var/lib/apt/lists/*

RUN export DEBIAN_FRONTEND=noninteractive \
    && gpg --keyserver pgpkeys.mit.edu --recv-key 5C808C2B65558117 \
    && gpg -a --export 5C808C2B65558117 | apt-key add - \
    && echo "deb http://www.deb-multimedia.org jessie main" > /etc/apt/sources.list.d/deb-multimedia.list \
    && apt-get update \
    && apt-get install -y deb-multimedia-keyring ffmpeg \
    && which ffmpeg \
    && rm -rf /var/lib/apt/lists/*

RUN mix local.hex --force

CMD ["iex"]
```

For the db we are using `postgres:9.5.3` and for asset compilation we are using `node:6.3.0`. This is the order in which we use scripts from `script`:

1. `script/deps_install` (`--only prod` not required in this case)
1. `script/assets_compile`
1. `script/assets_digest` (not applicable for non-prod)
1. `script/run`

**Comment by gerhard** [comment]
@jerodsanto I am keen on capturing any data seeding step in a script if possible. I believe that keeping all app concerns self-contained is a worthwhile effort.

File: config/test.exs [hunk 1]
```diff
@@ -16,7 +16,9 @@ config :changelog, Changelog.Mailer,
 # Configure your database
 config :changelog, Changelog.Repo,
   adapter: Ecto.Adapters.Postgres,
-  database: "changelog_test",
+  url: (
+    System.get_env("DB_URL") || "postgres://postgres@localhost:5432/changelog"
```
Comment (sobolevn) [inline config/test.exs]: How do you like this way of providing settings for both docker and normal development?

File: .dockerignore [hunk 2]
```diff
@@ -1,13 +1,11 @@
-
-/_build
-/db
-/deps
+_build
```
Comment (sobolevn) [inline .dockerignore]: This is required to make things work: https://github.com/docker/compose/issues/1607

File: config/test.exs (hunk 1 above)
Comment (jerodsanto) [inline config/test.exs]: Seems fine but why append `"_test"` to the end? Why not just have that be set in the environment variable?

File: script/run [hunk 3]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash -ex
 
-mix phoenix.server
+DB_URL="postgres://postgres@localhost:5432/changelog_dev" mix phoenix.server
```
Comment (jerodsanto) [inline script/run]: I'm pretty sure this would jack up our production deploy. @gerhard do you agree?

Comment (gerhard) [inline script/run]: Yes, definitely would affect production.

**Comment by sobolevn** [comment]
Sorry, the work took me long to finish this.

**Comment by jerodsanto** [comment]
This PR has been replaced by #246, which just got merged. Thanks for getting the ball rollin', @sobolevn! 💚💚💚

**Pull request closed without merging**